multi.py

```diff
@@ -0,0 +1,134 @@
+        process_args: Tuple of (process_id, query_id, results, bench_kwargs)
+    """
+    process_id, query_id, bench_kwargs = process_args
+    try:
+        # Modify prefix for this process group
+        bench_kwargs = bench_kwargs.copy()
+        bench_kwargs["prefix"] = f"{bench_kwargs['prefix']}-{process_id}"
+
+        benchmark = Benchmark(**bench_kwargs)
+        result = benchmark.run()
+        return result.to_json()
+
+    except Exception as e:
+        print(f"Process {process_id}, query {query_id} failed: {e}", file=sys.stderr)
+        print(traceback.format_exc(), file=sys.stderr)
+        return None
+
+
+def run_multi_benchmark(
+    num_processes: int,
+    query_process: int,
+    dataset: str,
+    num_tables: int,
+    batch_size: int,
+    num_queries: int,
+    ingest: bool,
+    index: bool,
+    prefix: str,
+    reset: bool,
+) -> BenchmarkResults:
+    total_processes = num_processes * (query_process if not ingest and not index else 1)
+    print(f"Starting {total_processes} benchmark processes...")
+
+    bench_kwargs = {
+        "dataset": dataset,
+        "num_tables": num_tables,
+        "batch_size": batch_size,
+        "num_queries": num_queries,
+        "ingest": ingest,
+        "index": index,
+        "prefix": prefix,  # Base prefix, will be modified per process
+        "reset": reset,
+    }
+
+    process_args = []
+
+    if ingest or index:
+        for i in range(0, num_processes):
+            process_kwargs = bench_kwargs.copy()
+            process_args.append((i + 1, 0, process_kwargs))
+    else:
+        for i in range(0, num_processes):
+            for j in range(0, query_process):
+                process_kwargs = bench_kwargs.copy()
+                process_args.append((i + 1, j, process_kwargs))
+
+    with mp.Pool(processes=total_processes) as pool:
+        process_results = pool.map(run_benchmark_process, process_args)
+
+        successful_results = [
+            BenchmarkResults.from_json(r) for r in process_results if r is not None
+        ]
+
+        if not successful_results:
+            raise RuntimeError(
+                "All benchmark processes failed - check logs for details"
+            )
+
+        return BenchmarkResults.combine(successful_results)
+
+
+def validate_args(args: argparse.Namespace):
+    if args.query_process > 1:
+        if args.ingest or args.index:
+            raise ValueError(
+                "Multiple query processes per table (query_process > 1) is only allowed "
+                "with --no-ingest and --no-index flags"
+            )
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument
+    parser.add_argument(
+        "-n",
+        "--num_processes",
+        type=int,
+        required=False,
+        default=1,
+        help="Number of total benchmark process. This number should be the same for data ingestion and data querying.",
+    )
+    parser.add_argument(
+        "-qn",
+        "--query_process",
```

Comment:
Why separate parameters for querying and ingestion? This might be confusing for users

Reply:
This is because our previous script is optimized for total QPS. Our total QPS across all tables are very good, but per table QPS is around 18 which harvey is asking for 50 or even more.

having multiple process querying a table can probably give us best per table QPS perf.

Any suggestion for that? should i change to use multiple threads querying a table? but threads in python is not actually multi-threading. Haven't done the tests, not quite sure about the overall perf of this other approach

```
Single table concurrent read test
uv run multi.py -n 1 -t 1 -p table_prefix -q 100 -r
create 1 process, each create 1 tables with 1 million rows with table name with table_prefix
for ingestion and indexing
and then run
uv run multi.py -n 1 -t 1 -p table_prefix -q 10000 -qn 10 -no-ingest --no-index
to run 10 processes (-qn query process number) to each run 10000 queries against the created single table (edited)
```